pinc/unicode.inc

```diff
@@ -18,47 +18,64 @@ function utf8_normalize($string)
 # Filter to only characters in $string that are in $valid_codepoints.
 function utf8_filter_to_codepoints($string, $valid_codepoints, $replacement="")
 {
-    $char_class = "";
-    foreach($valid_codepoints as $codepoint)
-    {
-        if(strpos($codepoint, '-') === False)
-        {
-            $char_class .= utf8_chr($codepoint);
-        }
-        else
+    $pattern_string = make_regex_string($valid_codepoints);
+    $pattern_string = "/$pattern_string/u";
+    return preg_replace_callback(
+        "/\X/u",
+        function ($matches) use($pattern_string)
         {
-            list($start, $end) = explode('-', $codepoint);
-            $char_class .= utf8_chr($start) . '-' . utf8_chr($end);
-        }
-    }
-    if(!$char_class)
-        return $string;
-
-    $pattern_string = "/[^$char_class]/u";
-    return preg_replace($pattern_string, $replacement, $string);
+            $match = $matches[0];
+            if(1 === preg_match($pattern_string, $match))
+            {
+                return $match;
+            }
+            return "";
+        },
+        $string);
 }
 
 # Filter out any characters in $string that are in $remove_codepoints
 function utf8_filter_out_codepoints($string, $remove_codepoints, $replacement="")
```

Comment:
I don't understand why this function needs to change. The purpose of this function is to remove specific codepoints from a string. If, instead, we want to remove specific characters, which might include combined characters, that would be a different function with a different name.

Reply:
You are right, it doesn't. It was formerly used in `get_invalid_characters()` but I cghanged that. I've put back the original.